client/beefy/src/worker.rs

```diff
@@ -811,6 +771,22 @@ where
 		);
 
 		loop {
+			// Don't bother voting or requesting justifications during major sync.
+			if !self.network.is_major_syncing() {
+				// If the current target is a mandatory block,
+				// make sure there's also an on-demand justification request out for it.
+				if let Some(block) = self.voting_oracle().mandatory_pending() {
+					// This only starts new request if there isn't already an active one.
+					self.on_demand_justifications.request(block);
+				}
+				// There were external events, 'state' is changed, author a vote if needed/possible.
+				if let Err(err) = self.try_to_vote() {
```

Comment:
Just thinking loud - not an issue or comment regarding the code. So if I get it correctly, the state is saved only when round is concluded.

... then if current round has not yet concluded and the validator has restarted:
1) it may e.g. cast a double vote - the fact that the vote has been casted already is not saved;
2) it'll lose all incoming votes - in worst case the round may not conclude at all.
That is fine until we have no any penalties for double votes and etc.

... then if current round has already finished and the validator has restarted after that:
1) it'll cast a (potentially second) vote for some obsolete round;
2) it'll reject all incoming votes for new round until he'll see a BEEFY finality.
That is fine until we have no any penalties for double votes and etc.

But if at some point we'll be penalizing validators for their incorrect behavior, it'll need to be fixed in some way. E.g. to save fact of broadcasting self vote + probably looking at the current round before voting (it may break something though).

Reply:
I was thinking to handle this when impl equivocation, but after running your good scenario breakdown in my head, I think we can easily "fix" it in this PR and cover all cases. Let me know if the following makes sense:

#### A. round has not yet concluded and the validator has restarted:
> it may e.g. cast a double vote - the fact that the vote has been casted already is not saved;

correct, let's fix it by saving full state on self vote.

> it'll lose all incoming votes - in worst case the round may not conclude at all

also correct, but it shouldn't be a problem:
- if less than 1/3 validators are restarted, they will simply conclude whatever (future) round the others are voting on and continue from there,
- if more than 1/3 validators are restarted in some weird cascade where they lost enough votes, worst case is they are stuck there until next BEEFY mandatory is GRANDPA finalized and they move to try to finalize that one

is in fact a problem if >1/3 are restarted while running a mandatory round -> we should also persist incoming votes if they're for mandatory rounds

#### B. validator restarted at an old concluded round

> it'll cast a (potentially second) vote for some obsolete round

all good: we fix double voting by saving state on vote, and voting on obsolete rounds is not misbehavior

> it'll reject all incoming votes for new round until he'll see a BEEFY finality

all good: it'll not reject votes (unless obsolete), it processes votes for any block/round within the `VoterOracle::accepted_interval()` (which is any block within current session if mandatory one is done), drop if `vote < interval` (obsolete) and enqueue for future processing if `vote > interval`.
Also voter cannot get stuck on a mandatory block that has been finalized while it was offline/restarting because of the on-demand-justification mechanism which greedily requests justifications for mandatory blocks -> some other peer will provide a justification for it if it's been finalized.

Reply:
Let's probably leave as is and leave it for equivocation PR :) But it is up to you :)